src/ray/object_manager/test/pull_manager_test.cc

```diff
@@ -127,26 +135,69 @@ TEST_F(PullManagerTest, TestRetryTimer) {
 
   for (; fake_time_ <= 127 * 10; fake_time_ += 1.) {
     pull_manager_.Tick();
-  }
-
-  // Rapid set of location changes.
-  for (int i = 0; i < 127; i++) {
-    fake_time_ += 0.1;
+    // This OnLocationChange shouldn't trigger an additional pull request or update the
+    // retry time.
     pull_manager_.OnLocationChange(obj1, client_ids, "");
   }
 
-  // We should make a pull request every tick (even if it's a duplicate to a node we're
-  // already pulling from).
-  // OnLocationChange also doesn't count towards the retry timer.
-  // To the casual observer, this may seem off-by-one, but this is due to floating point
-  // error (0.1 + 0.1 ... 10k times > 10 == True)
-  ASSERT_EQ(num_send_pull_request_calls_, 1 + 7 + 127);
+  ASSERT_EQ(num_send_pull_request_calls_, 1 + 7);
   ASSERT_EQ(num_restore_spilled_object_calls_, 0);
 
   pull_manager_.CancelPull(obj1);
   ASSERT_EQ(pull_manager_.NumActiveRequests(), 0);
 }
 
+TEST_F(PullManagerTest, TestDedupe) {
+  ObjectID obj1 = ObjectID::FromRandom();
+  rpc::Address addr1;
+  ASSERT_EQ(pull_manager_.NumActiveRequests(), 0);
+  pull_manager_.Pull(obj1, addr1);
+  ASSERT_EQ(pull_manager_.NumActiveRequests(), 1);
+
+  std::unordered_set<NodeID> pulled_nodes;
+
+  std::unordered_set<NodeID> client_ids;
+  client_ids.insert(NodeID::FromRandom());
+  pull_manager_.OnLocationChange(obj1, client_ids, "");
+  pulled_nodes.insert(last_pulled_node_);
+
+  ASSERT_EQ(num_send_pull_request_calls_, 1);
+  ASSERT_EQ(num_restore_spilled_object_calls_, 0);
+
+  // 2 additional nodes have the object. We don't pull it yet though, because the retry
+  // timer hasn't gone off yet.
+  client_ids.insert(NodeID::FromRandom());
+  client_ids.insert(NodeID::FromRandom());
+  pull_manager_.OnLocationChange(obj1, client_ids, "");
+
+  ASSERT_EQ(num_send_pull_request_calls_, 1);
+  ASSERT_EQ(num_restore_spilled_object_calls_, 0);
+
+  fake_time_ += 10;
+  pull_manager_.Tick();
+  pulled_nodes.insert(last_pulled_node_);
+
+  fake_time_ += 20;
+  pull_manager_.Tick();
+  pulled_nodes.insert(last_pulled_node_);
+
+  // Now that 2 retry intervals have passed, we should've pulled from both of the other
+  // nodes.
+  ASSERT_EQ(num_send_pull_request_calls_, 3);
+  ASSERT_EQ(num_restore_spilled_object_calls_, 0);
+  ASSERT_EQ(pulled_nodes.size(), 3);
+
+  fake_time_ += 40;
+  pull_manager_.Tick();
+  pulled_nodes.insert(last_pulled_node_);
+
+  // Since we've sent a pull to every node we know about, we should now retry from a node
+  // we've already pulled from.
+  ASSERT_EQ(num_send_pull_request_calls_, 4);
+  ASSERT_EQ(num_restore_spilled_object_calls_, 0);
+  ASSERT_EQ(pulled_nodes.size(), 3);
```

Comment:
It would be good to assert that the pulled_nodes' contents are the same as the client locations, and that each client was tried the same number of times after the first and second rounds.